File: tests/e2e/cases/06-int-ops/is_power_of_two/is_power_of_two_verified.cpp
/*
DEMO: Power-of-two predicate via the classic `x & (x - 1) == 0` bit trick.

  * Returns 1 when x is exactly 2^k for 0 <= k < 32,
  * Returns 0 otherwise.
  * Notably handles x == 0 correctly (the bit trick alone would say
    "yes" because 0 & (0 - 1) == 0).

Coverage: bit-trick predicate over a single uint32 with a 0/1-valued
return — pattern absent from earlier tests.

Expected verdict: VERIFIED.
*/

#include <cstdint>

uint32_t is_power_of_two(uint32_t x) {
    return (x != 0 && (x & (x - 1)) == 0) ? 1u : 0u;
}

File: tests/e2e/cases/06-int-ops/is_power_of_two/is_power_of_two_verified.rs
// DEMO: Power-of-two predicate (Rust mirror of is_power_of_two_verified.cpp).
//
// Returns 1 iff x is exactly 2^k for 0 <= k < 32. The `wrapping_sub`
// keeps the `x = 0` case well-defined (0u32.wrapping_sub(1) = 0xFFFFFFFF),
// then the `x != 0` guard rejects it.

fn is_power_of_two(x: u32) -> u32 {
    if x != 0 && (x & x.wrapping_sub(1)) == 0 { 1 } else { 0 }
}
